### PR TITLE
Add support for serializing debug_value instructions

### PR DESCRIPTION
**Pull request by usama54321**
This PR adds support for serialization and deserialization of debug value instructions

### CODE REVIEW

**Comment by usama54321** [comment]
Need to write tests for the initial commit + 4 bugfixes

**Comment by usama54321** [comment]
@swift-ci test

**Comment by usama54321** [comment]
@swift-ci test

File: lib/IRGen/IRGenSIL.cpp [hunk 1]
```diff
@@ -965,6 +965,9 @@ class IRGenSILFunction :
   bool shouldShadowVariable(SILDebugVariable varInfo, bool isAnonymous) {
     return !IGM.IRGen.Opts.DisableDebuggerShadowCopies
       && !IGM.IRGen.Opts.shouldOptimize()
+      //Disable shadow in deserialized functions as they might be optimized
```
Comment (adrian-prantl) [inline lib/IRGen/IRGenSIL.cpp:968]:
```suggestion
      // Shadow copies are only emitted at -Onone, but a deserialized function might have been already optimized, so ignore those.
```

File: lib/SIL/IR/SILBuilder.cpp [hunk 2]
```diff
@@ -638,17 +638,24 @@ DebugValueInst *SILBuilder::createDebugValue(SILLocation Loc, SILValue src,
...
 
-  // Debug location overrides cannot apply to debug value instructions.
-  DebugLocOverrideRAII LocOverride{*this, std::nullopt};
-  return insert(DebugValueInst::create(
-      getSILDebugLocation(Loc, true), src, getModule(),
-      *substituteAnonymousArgs(Name, Var, Loc), poisonRefs, moved, trace));
+  SILDebugLocation DebugLoc;
+  if (overrideLoc) {
+    // Debug location overrides cannot apply to debug value instructions.
+    DebugLocOverrideRAII LocOverride{*this, std::nullopt};
```
Comment (adrian-prantl) [inline lib/SIL/IR/SILBuilder.cpp:650]: Does putting this into the `{` block mean that it won't affect the new instruction? Is that what you want here?

File: lib/Serialization/SILFormat.h [hunk 3]
```diff
@@ -312,6 +314,19 @@ namespace sil_block {
     SILTypeCategoryField 
   >;
 
+  using SILDebugValueLayout = BCRecordLayout<
+    SIL_DEBUG_VALUE,
+    
+    SILTypeCategoryField, // operand type category
+    SILTypeCategoryField, // debug var type category
+    BCFixed<11>,// poison, movableValueDebuginfo, trace, hasDebugVar, isLet, isDenseMapSingleton(two bits), hasSource, hasLoc, hasExpr
```
Comment (adrian-prantl) [inline lib/Serialization/SILFormat.h:322]: clang-format

File: lib/SIL/IR/SILBuilder.cpp (hunk 2 above)
Comment (usama54321) [reply lib/SIL/IR/SILBuilder.cpp:650]: The loc override applied in the getSILDebugLocation below. I am keeping the old behavior consistent while allowing us to not override the loc in the else statement below while deserializing

**Comment by usama54321** [comment]
@swift-ci test please

**Comment by usama54321** [comment]
@swift-ci test

**Comment by usama54321** [comment]
@swift-ci test

File: lib/Serialization/DeserializeSIL.cpp [hunk 4]
```diff
@@ -430,9 +430,7 @@ SILDeserializer::readLoc(unsigned kind, SmallVectorImpl<uint64_t> &scratch) {
   case SILLocation::RegularKind:
     return RegularLocation(FNameLoc, Implicit);
   }
-
-  // LocationKind was not a recognized SILLocation::LocationKind.
-  return std::nullopt;
+  llvm_unreachable("Invalid LocationKind");
```
Comment (adrian-prantl) [inline lib/Serialization/DeserializeSIL.cpp:433]: We should not use llvm_unreachable to deal with invalid input. Code like this can crash long-running tools like LLDB which makes for a very unpleasant user experience. Can you just return an error instead?

File: lib/Serialization/SerializeSIL.cpp [hunk 5]
```diff
@@ -974,7 +978,108 @@ void SILSerializer::writeSILInstruction(const SILInstruction &SI) {
...
+          attrs |= 1 << 9;
+          auto LC = SM.getPresumedLineAndColumnForLoc(Loc);
+          auto FName = SM.getDisplayNameForLoc(Loc);
+          auto FNameID = S.addUniquedStringRef(FName);
+
+          ListOfValues.push_back(LC.first);
+          ListOfValues.push_back(LC.second);
+          ListOfValues.push_back(FNameID);
+        } else if (RawLoc.isFilenameAndLocation()) {
+          // TODO: this is a workaround until rdar://problem/25225083 is
```
Comment (adrian-prantl) [inline lib/Serialization/SerializeSIL.cpp:1034]: This comment isn't useful for Swift contributors without access to the link. Can you summarize what the problem is here?

File: lib/Serialization/SerializeSIL.cpp [hunk 6]
```diff
@@ -974,7 +978,108 @@ void SILSerializer::writeSILInstruction(const SILInstruction &SI) {
...
+    auto &SM = SI.getFunction()->getModule().getSourceManager();
+
+    SmallVector<uint64_t, 8> ListOfValues;
+    ListOfValues.push_back(addValueRef(Operand));
+    ListOfValues.push_back(S.addTypeRef(Type.getRawASTType()));
+
+    if (DebugVar) {
+      attrs |= 1 << 3; // has debug var
+      attrs |= DebugVar->isLet() << 4;
+      attrs |= DebugVar->isDenseMapSingleton << 5; // needs two bits
```
Comment (adrian-prantl) [inline lib/Serialization/SerializeSIL.cpp:1005]: Can you adjust these comments to the LLVM/Swift coding style? (full sentences)

**Comment by usama54321** [comment]
@swift-ci test

**Review by adrian-prantl — APPROVED** [review]
